### ISSUE TITLE
to_dict() should take an options object or pass through kwargs

### ISSUE THREAD

**Issue by mstump** [Fatal1ty/mashumaro#1]
-  I want to interact with outside data types that need ```camelKeys```, but want to use ```snake_keys``` for Python objects.
- I want these rules to be applied recursively so that whenever we convert a child ```DataClassDictMixin```the behavior is carries through.

I could get most of this behavior by overloading to_dict, and intercepting the input or output. The problem is that in some scenarios I want the camelCase behavior and in some scenarios I don't. An example would be JSON serialization where it's enabled, and dict serialization where it's not. If I add a flag to ```to_dict``` to control the behavior it won't pass through to nested instances. We could convert the method to take an Options dict, or we could pass through kwargs to recursive invocations.

**Comment by mstump** [comment]
Hmm, this won't work because it looks like we overwrite the to_dict with the generated method so it can't be overloaded. Encode and decode hooks that are passed through instead?

**Comment by mstump** [comment]
I'm solving two problems with the hook. 

- The first is the snake and camel conversion. We've got a type that has a user provided map. We'd want to camel the top level keys but leave the end-user supplied keys alone. 
- The second is that the dataclass being converted have fields that use variable encoding and for performance reasons aren't decoded/cached until requested through a property. The encoding information is kept as an enum on the dataclass that I uses internally but wouldn't want to be exposed via JSON.

In my hook I'm camel/snake casing the top level keys, but leaving the child keys. 
I'm getting the value of a property and inserting it into the dict
I'm stripping enum values from the dataclass.

The main difference between my hook and your encoder is that my hook retains the context in which the conversion code is being called. On encode I've got the dict that library produced, and the instance that is being converted, on decode I've got the input dict and the target class it allows me to store data on the instance/class and use that to make runtime decisions.

In the example below I'm dropping fields and replacing the content of one field with decoded data from the property.

Because the type conversion decisions are done at class creation, what I'm doing doesn't alleviate the need for a type registry but it does let me do handle a pretty broad/flexible set of problems. In an ideal world I think I'd want the ability to add field level annotations that let me control the encode/decode behavior per field but that's not possible in Python. The equivalent would be a map of field or type names to functions/hooks in addition to this dataclass level hook to tackle everything. Essentially for each field follow the field specific behavior, and after that's done apply X transform.

```
def opaque_encodable_encode_hook(instance, out_dict):
    if isinstance(instance, OpaqueEncodableFieldMixin):
        for f in instance.__class__.opaque_encodable_fields():
            out_dict.update({f.raw_field_name: out_dict.get(f.property_name, None)})
            out_dict.pop(f.property_name, None)

    return out_dict


def opaque_encodable_decode_hook(cls, input_dict):
    if issubclass(cls, OpaqueEncodableFieldMixin):
        for f in cls.opaque_encodable_fields():
            input_dict.update({f.cache_attribute_name: input_dict.get(f.raw_field_name, None)})
            input_dict.pop(f.raw_field_name, None)

    return input_dict

class JSONEncodableMixin(DataClassDictMixin):
    JSONEncodeHook = Callable[[Any, dict], dict]
    JSONDecodeHook = Callable[[Type[Any], dict], dict]

    def __init_subclass__(
            cls,
            json_encode_hook: Optional[JSONEncodeHook] = None,
            json_decode_hook: Optional[JSONDecodeHook] = None,
            json_register_faust_serializer=False,
            **kwargs) -> None:

        super().__init_subclass__(**kwargs)
        cls.__json_encode_hook = json_encode_hook
        cls.__json_decode_hook = json_decode_hook

    def to_json(
            self,
            camel_keys=True):

        def hook(instance, out_dict):
            if self.__class__.__json_encode_hook:
                out_dict = self.__class__.__json_encode_hook(instance, out_dict)

            if camel_keys:
                out_dict = camel_dict(out_dict)

            return out_dict

        return rapidjson.dumps(self.to_dict(encode_hook=hook))
```

**Comment by mstump** [comment]
Hey, just an FYI. I ended up going a different route and wrote something from scratch. I don't need this PR. 

I ended up using a combination of ideas. Instead of emitting code I create a map in the sub-class init that which maps field names to codecs. The class init hook goes through the fields combining the type information with the user supplied input:

* optional partial type->codec map
* optional partial field->codec map
* optional partial property->codec map
* pre and post hooks for both encode and decode
* list of fields to omit

When encode or decode is called instead of doing introspection like most libraries I just iterate through the map pulling the codec for each field, convert and return the output. Like you I avoid the runtime type checking cost on each conversion, but I use a map instead of emitting code.
